### PR TITLE
Implement and Test Matrix and Vector Classes

### PR DESCRIPTION
**Pull request by GigaKekes**
This pull request introduces two new classes, Matrix and Vector, that support basic mathematical operations. Both classes offer functionality that integrates well with NumPy and Python's standard indexing.

## 1. Matrix Class Implementation:

- `__init__`: Initializes the matrix using a 2D iterable.
- `__add__`: Adds two matrices.
- `__matmul__`: Multiplies two matrices.
- `T`: Returns the transpose of the matrix.
- `__getitem__`: Allows access to matrix elements using mat[i, j].
- `__repr__ `and `__str__`: Provides string representations of the matrix for easy reading and debugging.
- `__array__`: Enables interaction with NumPy methods directly on the matrix object.
- `data`: Property to access the underlying matrix.

## 2. Vector Class Implementation:

- `__init__`: Initializes the vector using an iterable.
- `__array__`: Allows the Vector object to be treated as a NumPy array directly.
- `__getitem__`: Enables access to elements using vector indexing.
- `__len__`: Returns the length (dimension) of the vector.
- `__mul__`: Calculates the dot product of two vectors.
- `__xor__`: Computes the angle (in radians) between two vectors.
- `norm`: Calculates the Euclidean norm (magnitude) of the vector.

## Tests

- Matrix Class:
- - Unit tests for matrix addition, multiplication, transpose, item access, and string representations.
- - Tests for invalid operations (e.g., shape mismatches during matrix multiplication).
- Vector Class:
- - Unit tests for vector operations such as dot product, angle calculation, norm, and element access.
- - Compatibility tests with NumPy array operations using __array__.

### CODE REVIEW

File: project/vector_matrix_operations/Matrix.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        Returns the result of matrix addition with another matrix.
+
+    __matmul__(other: "Matrix") -> "Matrix"
+        Returns the result of matrix multiplication with another matrix.
+
+    T() -> "Matrix"
+        Returns the transpose of the matrix.
+    """
+
+    def __init__(self, object):
```
Comment (Krekep) [inline project/vector_matrix_operations/Matrix.py]: У `object` отсутствует typehint

File: project/vector_matrix_operations/Vector.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        Returns the dot product of the current vector with another vector.
+
+    __xor__(other: "Vector") -> float
+        Returns the angle (in radians) between the current vector and another vector.
+
+    norm() -> float
+        Returns the norm (magnitude) of the vector.
+    """
+
+    def __init__(self, object):
```
Comment (Krekep) [inline project/vector_matrix_operations/Vector.py]: У `object` отсутствует typehint

File: tests/test_vector_matrix_operations/test_matrix_operations.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    ), "Matrix multiplication with an identity matrix is incorrect"
+
+
+def test_matrix_multiplication_shape_mismatch():
+    mat1 = Matrix([[1, 2], [3, 4]])
+    mat2 = Matrix([[5, 6, 7]])
+    with pytest.raises(
+        ValueError, match="Matrix shapes are not compatible for multiplication"
+    ):
+        mat1 @ mat2
```
Comment (Krekep) [inline tests/test_vector_matrix_operations/test_matrix_operations.py]: Не хватает теста на умножение не квадратных матриц

**Review by Krekep — CHANGES_REQUESTED** [review]
В целом хорошо, но необходимы исправления.
По поводу typehints у `object` --- либо сузить область допустимых значений, либо посмотреть в сторону collections.abc.Iterable, но я с этим не работал, так что не уверен, что получится.

**Comment by GigaKekes** [comment]
### Added
- Typehint for parameter `object` in classes `Vector` и `Matrix`
- Python version in `.yml`  files were updated to 3.12.2 so `X | Y` syntax for unions can be used
### Tests
- Test on rectangular matrix multiplication  `test_matrix_multiplication_non_square`

**Review by Krekep — APPROVED** [review]
Зачтено

**Pull request merged**